### PR TITLE
Fix Jest Matchers type

### PR DESCRIPTION
**Pull request by mattphillips**
This PR addresses the `Matchers<R>` -> `Matchers<R, T>` change in #39243

The above change is not inline with the types of `Jest` [see code here](https://github.com/facebook/jest/blob/master/packages/expect/src/types.ts#L87) and has broken any libraries extending Jest's `expect` (see: https://github.com/jest-community/jest-extended/issues/234).

The change in this PR adds a default to the new `T` type parameter of `void` which means that other libraries that extend `expect` can write their matchers in terms of `Matchers<R>` or `Matchers<R, T>` (should they need the `T`).

cc/ @tonyhallett

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [ ] Test the change in your own code. (Compile and run.)
- [ ] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [ ] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/facebook/jest/blob/master/packages/expect/src/types.ts#L87

### CODE REVIEW

**Review by wsmd — APPROVED** [review]
Thanks for this @mattphillips!

File: types/jest/index.d.ts [hunk 1]
```diff
@@ -727,7 +727,7 @@ declare namespace jest {
         not: T
     };
     // should be R extends void|Promise<void> but getting dtslint error
-    interface Matchers<R, T> {
+    interface Matchers<R, T = void> {
```
Comment (uniqueiniquity) [inline types/jest/index.d.ts]: It seems like `T` isn't used in return type positions, but `void` only really makes sense as the return type of a function.
If your intent is to have a default with no properties, could you use `{}` here instead? Thanks!

**Review by uniqueiniquity — CHANGES_REQUESTED** [review]

File: types/jest/index.d.ts (hunk 1 above)
Comment (wsmd) [reply types/jest/index.d.ts]: Good catch @uniqueiniquity! 

Using `void` is likely to cause issues with `toMatchSnapshot`/`toMatchInlineSnapshot` when `T` remains unspecified. Looks like we're expecting an indexed type here:

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/c640483291a20a1e7d3ebe1f13a444c697d7be4c/types/jest/index.d.ts#L977

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/c640483291a20a1e7d3ebe1f13a444c697d7be4c/types/jest/index.d.ts#L988

**Review by wsmd — CHANGES_REQUESTED** [review]

File: types/jest/index.d.ts (hunk 1 above)
Comment (SimenB) [reply types/jest/index.d.ts]: it was gonna be `= {}` but @orta suggested `= void` instead - maybe he has some thoughts he'd like to share?

Comment (orta) [reply types/jest/index.d.ts]: Hah, yeah, I recommended both - but I figured `void` would be closer to your average expect-y function type. Seems fine to me to use `{}` instead.

**Review by wsmd — APPROVED** [review]

**Review by uniqueiniquity — APPROVED** [review]

**Comment by uniqueiniquity** [comment]
Congratulations on your first DefinitelyTyped contribution!
Thank you for being a part of the community!

**Pull request merged**